File: ruby/path_traversal/file_download.rb
require 'sinatra'

BASE_DIR = '/var/www/files'

# VULN 1: File.read with string concatenation - no path sanitization
get '/read' do
  filename = params[:file]
  path = BASE_DIR + '/' + filename
  File.read(path)
end

# VULN 2: send_file with user-controlled path - allows ../../etc/passwd
get '/download' do
  filename = params[:file]
  filepath = File.join(BASE_DIR, filename)
  send_file filepath
end

# VULN 3: File.open with string interpolation - stream arbitrary files
get '/export' do
  report_name = params[:name]
  full_path = File.join('/reports/output', report_name)
  File.open(full_path, 'rb', &:read)
end

File: javascript/path_traversal/fileDownload.js
const express = require('express');
const fs = require('fs');
const path = require('path');

const router = express.Router();
const BASE_DIR = '/var/www/files';

// VULN 1: fs.readFile with string concatenation - no path sanitization
router.get('/read', (req, res) => {
    const filename = req.query.file;
    const filepath = BASE_DIR + '/' + filename;
    fs.readFile(filepath, 'utf8', (err, data) => {
        if (err) return res.status(500).send(err.message);
        res.send(data);
    });
});

// VULN 2: res.sendFile with user-controlled path - allows ../../etc/passwd
router.get('/download', (req, res) => {
    const filename = req.query.file;
    const filepath = path.join(BASE_DIR, filename);
    res.sendFile(filepath);
});

// VULN 3: fs.createReadStream with user-controlled path - stream arbitrary files
router.get('/export', (req, res) => {
    const reportName = req.query.name;
    const fullPath = path.join('/reports/output', reportName);
    const stream = fs.createReadStream(fullPath);
    stream.pipe(res);
});

module.exports = router;
